dm/config/source_config.go

```diff
@@ -353,3 +353,85 @@ func (c *SourceConfig) check(metaData *toml.MetaData, err error) error {
+func (c *SourceConfig) YamlForDowngrade() (string, error) {
+	s := NewSourceConfigForDowngrade(c)
+
+	// encrypt password
+	cipher, err := utils.Encrypt(utils.DecryptOrPlaintext(c.From.Password))
+	if err != nil {
+		return "", err
+	}
+	s.From.Password = cipher
+
+	// omit default values, so we can ignore them for later marshal
+	s.omitDefaultVals()
+
+	return s.Yaml()
+}
+
+// SourceConfigForDowngrade is the base configuration for source in v2.0.
+// This config is used for downgrade(export-configs) from a higher dmctl version.
+// When we add any new config item into SourceConfig, we should update it also.
+type SourceConfigForDowngrade struct {
+	EnableGTID      bool                   `yaml:"enable-gtid"`
+	AutoFixGTID     bool                   `yaml:"auto-fix-gtid"`
+	RelayDir        string                 `yaml:"relay-dir"`
+	MetaDir         string                 `yaml:"meta-dir"`
+	Flavor          string                 `yaml:"flavor"`
+	Charset         string                 `yaml:"charset"`
+	EnableRelay     bool                   `yaml:"enable-relay"`
+	RelayBinLogName string                 `yaml:"relay-binlog-name"`
+	RelayBinlogGTID string                 `yaml:"relay-binlog-gtid"`
+	UUIDSuffix      int                    `yaml:"-"`
+	SourceID        string                 `yaml:"source-id"`
+	From            DBConfig               `yaml:"from"`
+	Purge           PurgeConfig            `yaml:"purge"`
+	Checker         CheckerConfig          `yaml:"checker"`
+	ServerID        uint32                 `yaml:"server-id"`
+	Tracer          map[string]interface{} `yaml:"tracer"`
+	// any new config item, we mark it omitempty
+	CaseSensitive bool                  `yaml:"case-sensitive,omitempty"`
+	Filters       []*bf.BinlogEventRule `yaml:"filters,omitempty"`
+}
+
+// NewSourceConfigForDowngrade creates a new base config for downgrade.
+func NewSourceConfigForDowngrade(sourceCfg *SourceConfig) *SourceConfigForDowngrade {
+	return &SourceConfigForDowngrade{
+		EnableGTID:      sourceCfg.EnableGTID,
+		AutoFixGTID:     sourceCfg.AutoFixGTID,
+		RelayDir:        sourceCfg.RelayDir,
+		MetaDir:         sourceCfg.MetaDir,
+		Flavor:          sourceCfg.Flavor,
+		Charset:         sourceCfg.Charset,
+		EnableRelay:     sourceCfg.EnableRelay,
+		RelayBinLogName: sourceCfg.RelayBinLogName,
+		RelayBinlogGTID: sourceCfg.RelayBinlogGTID,
+		UUIDSuffix:      sourceCfg.UUIDSuffix,
+		SourceID:        sourceCfg.SourceID,
+		From:            sourceCfg.From,
+		Purge:           sourceCfg.Purge,
+		Checker:         sourceCfg.Checker,
+		ServerID:        sourceCfg.ServerID,
+		Tracer:          sourceCfg.Tracer,
+		CaseSensitive:   sourceCfg.CaseSensitive,
+		Filters:         sourceCfg.Filters,
+	}
+}
+
+// omitDefaultVals change default value to empty value for new config item.
+// If any default value for new config item is not empty(0 or false or nil),
+// we should change it to empty.
+func (c *SourceConfigForDowngrade) omitDefaultVals() {
```

Comment:
should this function be removed?

Reply:
I think we should keep it, for later use.